### PR TITLE
Fixed trace of Frobenius. Added tests.

### PR DESCRIPTION
**Pull request by JHanselman**
Also commented out some code that's no longer used, but I didn't want to delete quite yet.

### CODE REVIEW

File: src/EllCrv/Finite.jl [hunk 1]
```diff
@@ -807,9 +807,12 @@ function trace_of_frobenius(E::EllCrv{T}, n::Int) where T<:FinFieldElem
   a = q +1 - order(E)
   R, x = PolynomialRing(QQ)
   f = x^2 - a*x + q
-  L, alpha = NumberField(f)
-  rest = numerator(alpha^n + (-alpha + a)^n)
-  return q^n + 1 - rest
+  if isirreducible(f)
+    L, alpha = NumberField(f)
+  else
+    alpha = roots(f)[1]
```
Comment (thofma) [inline src/EllCrv/Finite.jl:813]: @JHanselman: Do you have an example where it uses the `else` clause?

Comment (JHanselman) [reply src/EllCrv/Finite.jl:813]: E = elliptic_curve_from_j_invariant(GF(2,4)(0))
trace_of_frobenius(E,2) == 32
trace_of_frobenius(E,3) == 128

should do it.

File: src/EllCrv/Finite.jl [hunk 2]
```diff
@@ -807,9 +807,12 @@ function trace_of_frobenius(E::EllCrv{T}, n::Int) where T<:FinFieldElem
   a = q +1 - order(E)
   R, x = PolynomialRing(QQ)
   f = x^2 - a*x + q
-  L, alpha = NumberField(f)
-  rest = numerator(alpha^n + (-alpha + a)^n)
-  return q^n + 1 - rest
+  if isirreducible(f)
+    L, alpha = NumberField(f)
```
Comment (thofma) [inline src/EllCrv/Finite.jl:811]:
```suggestion
    L, alpha = NumberField(f)
    return ZZ(trace(alpha^n))
```

**Comment by thofma** [comment]
There was no `trace` for rational numbers, so I refactored it a bit. Now the test is failing here: https://github.com/thofma/Hecke.jl/runs/6454810001?check_suite_focus=true#step:6:270

**Comment by JHanselman** [comment]
> There was no `trace` for rational numbers, so I refactored it a bit. Now the test is failing here: https://github.com/thofma/Hecke.jl/runs/6454810001?check_suite_focus=true#step:6:270

The output needs to be a^n+b^n where a and b are the two roots of the quadratic polynomial. This is identical to the trace when the polynomial is irreducible. For QQ it simply needs to be a^n+b^n.

File: src/EllCrv/Finite.jl [hunk 3]
```diff
@@ -807,9 +807,13 @@ function trace_of_frobenius(E::EllCrv{T}, n::Int) where T<:FinFieldElem
...
+    return ZZ(trace(alpha^n))
+  else
+    _alpha = roots(f)[1]
+    return ZZ(_alpha^n)
```
Comment (thofma) [inline src/EllCrv/Finite.jl:815]:
```suggestion
    _alpha, _beta = roots(f)
    return ZZ(_alpha^n + _beta^n)
```

**Comment by thofma** [comment]
OK, next try

File: src/EllCrv/Finite.jl [hunk 4]
```diff
@@ -807,9 +807,13 @@ function trace_of_frobenius(E::EllCrv{T}, n::Int) where T<:FinFieldElem
...
   f = x^2 - a*x + q
-  L, alpha = NumberField(f)
-  rest = numerator(alpha^n + (-alpha + a)^n)
-  return q^n + 1 - rest
+  if isirreducible(f)
+    L, alpha = NumberField(f)
+    return ZZ(trace(alpha^n))
+  else
+    _alpha, _beta = roots(f)
+    return ZZ(_alpha^n + _beta^n)
```
Comment (thofma) [inline src/EllCrv/Finite.jl:815]:
```suggestion
    _alpha = roots(f)[1]
    return 2 * ZZ(_alpha^n)
```

**Pull request merged**